File: LeetCode/1590/golang/lc1590.go
package lc1590

// Runtime complexity: O(n)
// Auxiliary space: O(n)
// Subjective level: medium/hard
// Solved on: 2025-11-30
func minSubarray(nums []int, p int) int {
	// WA #01: it's not allowed to remove the entire array!
	// so can't use an actual BIG_NUM, or rather when I tried this approach
	// the code could not register that the entire array is being deleted.
	//
	// const BIG_NUM = math.MaxInt32

	BIG_NUM := len(nums)
	sum := 0
	for _, num := range nums {
		sum += num
	}

	rem := sum % p
	if rem == 0 {
		return 0
	}

	prefixToIndex := map[int]int{}
	prefixToIndex[0] = -1
	prefix := 0
	ans := BIG_NUM

	for i, num := range nums {
		prefix += num
		prefix %= p
		target := (prefix - rem + p) % p
		val, keyExists := prefixToIndex[target]
		if keyExists {
			ans = min(ans, i-val)
		}
		prefixToIndex[prefix] = i
	}

	if ans == BIG_NUM {
		return -1
	}

	return ans
}

File: LeetCode/1590/python/lc1590.py
from typing import List

# Runtime complexity: O(n)
# Auxiliary space complexity: O(n)
# Subjective level: medium
# Solved on: 2026-02-19
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        L = len(nums)
        S = sum(nums)
        rem = S % p
        if rem == 0:
            return 0

        prefixSumModp = 0
        prefixSumToIndex = {0: -1} # maps prefix sum modulo p to array index.
        ans = L
        for i, num in enumerate(nums):
            prefixSumModp += num
            prefixSumModp %= p
            target = (prefixSumModp + p - rem) % p
            if target in prefixSumToIndex:
                ans = min(ans, i - prefixSumToIndex[target])
            prefixSumToIndex[prefixSumModp] = i

        return -1 if ans == L else ans
